src/Microsoft.AspNet.Mvc.Core/Formatters/AcceptHeaderParser.cs

```diff
@@ -0,0 +1,179 @@
+using System.Collections.Generic;
+using System.Diagnostics.Contracts;
+using System.Globalization;
+using Microsoft.Extensions.Primitives;
+
+namespace Microsoft.AspNet.Mvc.Formatters
+{
+    public static class AcceptHeaderParser
+    {
+        private static readonly StringSegment QualityParameter = new StringSegment("q");
+
+        public static IList<MediaTypeSegmentWithQuality> ParseAcceptHeader(IList<string> acceptHeaders)
+        {
+            var parsedValues = new List<MediaTypeSegmentWithQuality>();
+            ParseAcceptHeader(acceptHeaders, parsedValues);
+
+            return parsedValues;
+        }
+
+        public static void ParseAcceptHeader(IList<string> acceptHeaders, IList<MediaTypeSegmentWithQuality> parsedValues)
+        {
+            if (acceptHeaders == null)
+            {
+                throw new ArgumentNullException(nameof(acceptHeaders));
+            }
+
+            if (parsedValues == null)
+            {
+                throw new ArgumentNullException(nameof(parsedValues));
+            }
+
+            // If a parser returns an empty list, it means there was no value, but that's valid (e.g. "Accept: ").
+            // The caller can ignore the value.
+            if (acceptHeaders == null)
+            {
+                return;
+            }
+
+            foreach (var value in acceptHeaders)
+            {
+                int index = 0;
+
+                while (!string.IsNullOrEmpty(value) && index < value.Length)
+                {
+                    MediaTypeSegmentWithQuality output;
+                    if (TryParseValue(value, ref index, out output))
+                    {
+                        // The entry may not contain an actual value, like " , "
+                        if (output.MediaType.HasValue)
+                        {
+                            parsedValues.Add(output);
+                        }
+                    }
+                    else
+                    {
+                        throw new FormatException(string.Format(CultureInfo.InvariantCulture, "Invalid values '{0}'.",
+                            value.Substring(index)));
+                    }
+                }
+            }
+        }
+
+        private static bool TryParseValue(string value, ref int index, out MediaTypeSegmentWithQuality parsedValue)
+        {
+            parsedValue = default(MediaTypeSegmentWithQuality);
+
+            // If multiple values are supported (i.e. list of values), then accept an empty string: The header may
+            // be added multiple times to the request/response message. E.g.
+            //  Accept: text/xml; q=1
+            //  Accept:
+            //  Accept: text/plain; q=0.2
+            if (string.IsNullOrEmpty(value) || (index == value.Length))
+            {
+                return true;
+            }
+
+            var separatorFound = false;
+            var current = GetNextNonEmptyOrWhitespaceIndex(value, index, out separatorFound);
```

Comment:
No, current represents the current index while we are advancing and we don't want to pass index by ref as it's already by ref in that method and we don't want to change the index until we know for sure what situation we are in